src/cs50/sql.py

```diff
@@ -359,13 +381,15 @@ def shutdown_session(exception=None):
 
             # If constraint violated, return None
             except sqlalchemy.exc.IntegrityError as e:
+                self._disconnect()
                 self._logger.debug(termcolor.colored(statement, "yellow"))
                 e = RuntimeError(e.orig)
                 e.__cause__ = None
                 raise e
 
             # If user errror
             except sqlalchemy.exc.OperationalError as e:
+                self._disconnect()
```

Comment:
The `_disconnect()` calls in these exception handlers might be an issue since the behavior is a bit different in MySQL vs PostgreSQL. If e.g., a syntax error happens during a transaction, PostgreSQL automatically aborts the transaction while MySQL doesn't fyi.

Reply:
I think we're okay, though, since the connection is closed, which means any prior statements won't get committed anyway, thereby ensuring the whole transaction fails.

Reply:
Alright, no big deal. Just thought I'd bring up in case we want to be consistent with the MySQL client for example where we wouldn't disconnect on these errors and keep the transaction until the user commits or rolls back.

Reply:
Distinguishing between DB error and caller error now! https://github.com/cs50/python-cs50/pull/133/commits/31305e173a25366e45468af1f8767547f50e7e31